File: snippets/lesson-11-exhaustmap.ts
import { fromEvent, of } from 'rxjs';
import { exhaustMap, delay } from 'rxjs/operators';

const refreshBtn = document.getElementById('refresh-btn')!;

fromEvent(refreshBtn, 'click').pipe(
  exhaustMap(() => {
    console.log('Refreshing…');
    return of('Loaded').pipe(delay(2000));
  }),
).subscribe(status => console.log('Done:', status));

// Fast double-click:
// Refreshing…     ← first click
// Done: Loaded    ← 2s later
// (second click was ignored)

File: snippets/lesson-11-exhaustmap.cs
using System;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using System.Threading.Tasks;

var refreshes = new Subject<Unit>();

refreshes
    .Select(_ => Observable.FromAsync(async () =>
    {
        Console.WriteLine("Refreshing…");
        await Task.Delay(2000);
        return "Loaded";
    }))
    .Exhaust()
    .Subscribe(status => Console.WriteLine($"Done: {status}"));

refreshes.OnNext(Unit.Default); // first click
refreshes.OnNext(Unit.Default); // ignored while first is active
